pkg/targetgroupbinding/networking_manager.go

```diff
@@ -225,7 +230,103 @@ func (m *defaultNetworkingManager) consolidateIngressPermissionsPerSGByTGB(_ con
+					if _, ok := permsFromNLBPerSG[sgID]; !ok {
+						permsFromNLBPerSG[sgID] = []networking.IPPermissionInfo{}
+					}
+					permsFromNLBPerSG[sgID] = append(permsFromNLBPerSG[sgID], permission)
+				} else {
+					protocol := awssdk.StringValue(permission.Permission.IpProtocol)
+					if _, ok := permsByProtocolAndSourcePerSG[sgID][protocol]; !ok {
+						permsByProtocolAndSourcePerSG[sgID][protocol] = make(map[string][]networking.IPPermissionInfo)
+					}
+					groupID := ""
+					if len(permission.Permission.UserIdGroupPairs) == 1 {
+						groupID = awssdk.StringValue(permission.Permission.UserIdGroupPairs[0].GroupId)
+					}
+					if _, ok := permsByProtocolAndSourcePerSG[sgID][protocol][groupID]; !ok {
+						permsByProtocolAndSourcePerSG[sgID][protocol][groupID] = []networking.IPPermissionInfo{}
+					}
+					permsByProtocolAndSourcePerSG[sgID][protocol][groupID] = append(permsByProtocolAndSourcePerSG[sgID][protocol][groupID], permission)
+				}
+			}
+		}
+	}
+	return permsFromNLBPerSG, permsByProtocolAndSourcePerSG
+}
+
+// computeRestrictedIngressPermissionsPerSG will compute restricted ingress permissions group by source and protocol per SG
+func (m *defaultNetworkingManager) computeRestrictedIngressPermissionsPerSG(ctx context.Context) map[string][]networking.IPPermissionInfo {
+	permsFromNLBPerSG, permsByProtocolAndSourcePerSG := m.groupIngressPermsBySourceAndProtocolPerSG(ctx)
+
+	optimizedPermByProtocolAndSourcePerSG := make(map[string]map[string]map[string]networking.IPPermissionInfo)
+	for sgID, permsByProtocolAndSource := range permsByProtocolAndSourcePerSG {
+		if _, ok := optimizedPermByProtocolAndSourcePerSG[sgID]; !ok {
+			optimizedPermByProtocolAndSourcePerSG[sgID] = make(map[string]map[string]networking.IPPermissionInfo)
+		}
+		for protocol, permsBySource := range permsByProtocolAndSource {
+			if _, ok := optimizedPermByProtocolAndSourcePerSG[sgID][protocol]; !ok {
+				optimizedPermByProtocolAndSourcePerSG[sgID][protocol] = make(map[string]networking.IPPermissionInfo)
+			}
+			for groupID, perms := range permsBySource {
+				if _, ok := optimizedPermByProtocolAndSourcePerSG[sgID][protocol]; !ok {
+					optimizedPermByProtocolAndSourcePerSG[sgID][protocol][groupID] = networking.IPPermissionInfo{}
+				}
+				minPort, maxPort := defaultTgbMaxPort, defaultTgbMinPort
+				if len(perms) > 0 {
+					optimizedPermByProtocolAndSourcePerSG[sgID][protocol][groupID] = perms[0]
+				}
+				for _, perm := range perms {
+					if awssdk.Int64Value(perm.Permission.FromPort) > 0 && awssdk.Int64Value(perm.Permission.FromPort) < minPort {
+						minPort = *perm.Permission.FromPort
+					}
+					if awssdk.Int64Value(perm.Permission.ToPort) > maxPort {
+						maxPort = *perm.Permission.ToPort
+					}
+				}
+				if minPort > maxPort {
+					minPort, maxPort = defaultTgbMinPort, defaultTgbMaxPort
+				}
+				*optimizedPermByProtocolAndSourcePerSG[sgID][protocol][groupID].Permission.FromPort = *awssdk.Int64(minPort)
```

Comment:
imo, is is simpler to track the rules in a map[string][]networking.IPPermissionInfo? key is the sgID and the value is the list of permissions once we calculate the min and max ports.

Reply:
Done.